FIX fs: nvs: nvs fix static analysis error caused by mismatched variable assignment

andrisk-dev opened the pull request:
A variable of type size_t is assigned to a variable of type uint8_t in nvs_al_size function. This can trigger static analysis. The problem is fixed by using size_t instead of uint8_t.

Laczen commented:
@andrisk-dev thank you for the PR. NVS has been developped and is fit only for devices with a small write-block-size. Even a write-block-size of 32 is already wasting a lot of flash space and increasing it even more would make things even worse. 

If nvs is used to store e.g. 1 reboot counter of 32 bit on a device with 512B write-block-size would require at least 4kB for the nvs system, and each extra data item that needs storage would need an extra 2kB.

When using nvs with settings this would really be bad, to store 1 settings item at least 8kB is needed and each extra item would add 4kB.

These figures are so bad that they would only trigger complaints about the nvs system, and they can't be fixed. It seems more appropiate just to say that this is not supported. The littlefs system seems more appropiate to support these devices, while it also is not an ideal match with larger write-block-size it will be capable of better utilization of flash.

andrisk-dev committed:
fs: nvs: fix nvs static analysis error

There is asignment from size_t variable to uint8_t variable in
nvs_al_size function that can trigger static analysis error.
The problem is fixed by changing the variable to size_t type
instead of uint8_t.

Signed-off-by: Andrej Trnkóci <andrej.trnkoci@nxp.com>

Laczen commented:
Just as a comment, the figures are even worse than I mentioned. Each sector needs place for a close ate and space to allow a delete, so for this an extra 1kB will be required per nvs sector.

Laczen commented:
> Yes, it is obvious that using 512Byes write-block-size is inefficient. But we need to add this support. We should allow users to choose whether to use it or not.

I disagree, all but one of the in tree devices with write-block-size = <512> have very limited flash (268 kB - 512 kB), they don't have space for an inefficient storage solution.

In zephyr there also is support for bluetooth, but this won't work for devices without bluetooth hardware. This situation is similar, in zephyr there is nvs, but it doesn't support write-block-sizes larger than 32 byte. I would really like this limit not to be there but it is inherent to the design of nvs.

There is an "error" in nvs that it doesn't detect that write-block-size is truncated, but this is covered by write-block-size = 0. It could be a problem for devices with e.g. write-block-size=<513> but I have no knowledge of the existence of such devices.

butok commented:
> It could be a problem for devices with e.g. write-block-size=<513> but I have no knowledge of the existence of such devices.

There is no platforms with write-block-size=<513>.
Zephyr has platforms only with maximum  512Bytes.

> I disagree, all but one of the in tree devices with write-block-size = <512> have very limited flash (268 kB - 512 kB),
> they don't have space for an inefficient storage solution.

NXP has platforms with 2MB of FLASH and the 512 byte write!!!

Laczen commented:
> While I agree that the efficiency in this case is really bad I believe it can be improved a bit by using larger sector sizes. Then it is possible to get closer to 1kB per setting. For example when choosing sector size = 8kB, I believe 6 settings can fit to that space.

It might be worse than that. If you have only 2 sectors you can only fit 5 settings. On top of that once you have added these 5 settings you cannot change them without deleting them first. So in reality you can only use 4 settings.

I really cannot understand why you would like to enable such a clearly bad solution on any device.

butok commented:
> I really cannot understand why you would like to enable such a clearly bad solution on any device.

Even for a case with a few settings, this can make sense.
We should provide  tools, and an user project should choose what is appropriate and what is not.

Laczen commented:
> > I really cannot understand why you would like to enable such a clearly bad solution on any device.
> 
> Even for a case with a few settings, this can make sense. We should provide tools, and an user project should choose what is appropriate and what is not.

No it does not make sense. All this will generate is unhappy users complaining and there is nothing that can be done about it.

@andrisk-dev, please close this PR.

butok commented:
> > > I really cannot understand why you would like to enable such a clearly bad solution on any device.
> > 
> > 
> > Even for a case with a few settings, this can make sense. We should provide tools, and an user project should choose what is appropriate and what is not.
> 
> No it does not make sense. All this will generate is unhappy users complaining and there is nothing that can be done about it.
> 
> @andrisk-dev, please close this PR.

`const size_t write_block_size;`

BUT in wrong code
`uint8_t write_block_size = fs->flash_parameters->write_block_size;`
SO
uint8_t variable must be changed to size_t, in any case.

butok commented:
> > > > > I really cannot understand why you would like to enable such a clearly bad solution on any device.
> > > > 
> > > > 
> > > > Even for a case with a few settings, this can make sense. We should provide tools, and an user project should choose what is appropriate and what is not.
> > > 
> > > 
> > > No it does not make sense. All this will generate is unhappy users complaining and there is nothing that can be done about it.
> > > @andrisk-dev, please close this PR.
> > 
> > 
> > `const size_t write_block_size;`
> > BUT in wrong code `uint8_t write_block_size = fs->flash_parameters->write_block_size;` SO uint8_t variable must be changed to size_t, in any case.
> 
> There is no wrong code, it is already verified that write-block-size fits in a uint8_t.
> 
> The PR must be closed because the intention is wrong.
> 
> BTW adding reviewers from your own company to try and ~get what you want~ overrule a reviewers decision is bad practice. Select extra reviewers based on experience.

Anyway, uint32_t write-block-size may overflow uint8_t local variable.

de-nordic commented on subsys/fs/nvs/nvs.c:
```diff
@@ -88,7 +88,7 @@ static void nvs_lookup_cache_invalidate(struct nvs_fs *fs, uint32_t sector)
 static inline size_t nvs_al_size(struct nvs_fs *fs, size_t len)
 {
-	uint8_t write_block_size = fs->flash_parameters->write_block_size;
+	size_t write_block_size = fs->flash_parameters->write_block_size;
```
If there is concern that `write_block_size`, which has `size_t` type in the `struct flash_parameters` definition, can overflow the local definition, which is suitable to hold the design allowed values (https://github.com/zephyrproject-rtos/zephyr/pull/77569#issuecomment-2312766152), then `if` should be placed here, checking the value and reporting error; but I think that proper place for the warning would be in `nvs_mount`, not here.

butok replied:
The goal is to avoid the local variable overflow for all cases.

dleach02 replied:
The original code is going to get flagged by the static analysis due to the size mismatch of this assignment.

de-nordic replied:
> The original code is going to get flagged by the static analysis due to the size mismatch of this assignment.

That can be fixed with cast, but will not change design constrains.

> The goal is to avoid the local variable overflow for all cases.

Look, I do not have any solid arguments against using `size_t` here, so I guess the change itself is ok; but the fact that this can overflow comes from misuse where improper value got here without being vatted against the design constrains.

dleach02 replied:
understood. But lets move that discussion to the root you reference. 

We have raised an RFC #77576 for that discussion.

dleach02 commented:
> BTW adding reviewers from your own company to try and ~get what you want~ overrule a reviewers decision is bad practice. Select extra reviewers based on experience.

@Laczen please be aware that this is not the intent. 

You, as the assignee have overriding control over this PR within the boundaries called out in the Zephyr Project roles and responsibilities (https://docs.zephyrproject.org/latest/project/project_roles.html#assignee). Note that as the assignee, you have a responsibility to try to "_drive the pull request to a mergeable state_". If you disagree with the PR you can mark the PR with a change request which prevents it from being merged until the disagreements are resolved. Also, as the assignee, you are responsible for driving the [PR Technical Escalation process](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation). The first step is to discuss the issues in the developer working group.

Laczen commented:
A PR consists of:

a. A description of the change and why it is done as a commit message (and repeated as the start of the PR),
b. A code change that accompanies the PR,

When a PR is accepted both a. and b. are accepted. While I have no problem at all with the proposed code change, the
description and reason to do the code change are not acceptable for me as assignee of the PR. If the description and commit message are changed, and the link to the RFC is removed (while still allowing future eyes to understand the discussion) I see no reason not to merge the code change.

Instead of trying to squeeze this in the present PR it is simpler to create a new PR that only corrects nvs for a possible static analysis error.

butok commented:
It blocks our NXP LPC support,
so we have to escalate to upper levels, starting by the Zephyr: Dev. Meeting.

MaureenHelm commented:
> > @andrisk-dev thank you for the PR. NVS has been developped and is fit only for devices with a small write-block-size. Even a write-block-size of 32 is already wasting a lot of flash space and increasing it even more would make things even worse.
> > If nvs is used to store e.g. 1 reboot counter of 32 bit on a device with 512B write-block-size would require at least 2kB for the nvs system, and each extra data item that needs storage would need an extra 1kB.
> > When using nvs with settings this would really be bad, to store 1 settings item at least 4kB is needed and each extra item would add 2kB.
> > These figures are so bad that they would only trigger complaints about the nvs system, and they can't be fixed. It seems more appropiate just to say that this is not supported. The littlefs system seems more appropiate to support these devices, while it also is not an ideal match with larger write-block-size it will be capable of better utilization of flash.
> 
> Yes, it is obvious that using 512Byes write-block-size is inefficient. But we need to add this support. We should allow users to choose whether to use it or not.

Would like to understand why a user would want to use NVS with this inefficiency. What is the use case? I appreciate trying to make all subsystems work across all platforms, but this seems to be a case of a platform not supporting a subsystem's requirements.


> > BTW adding reviewers from your own company to try and ~get what you want~ overrule a reviewers decision is bad practice. Select extra reviewers based on experience.

Adding reviewers brings more stakeholders to the discussion.

> Also, as the assignee, you are responsible for driving the [PR Technical Escalation process](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-technical-escalation). The first step is to discuss the issues in the developer working group.

I misspoke in the dev-review meeting today when I said next step in the escalation process is the TSC. It's actually Architecture WG first, then TSC.

butok commented:
The PR was discussed on the dev-review meeting, so DEV label has been removed.

butok commented:
> Would like to understand why a user would want to use NVS with this inefficiency. What is the use case? I appreciate trying to make all subsystems work across all platforms, but this seems to be a case of a platform not supporting a subsystem's requirements.

Hi @MaureenHelm, 
For understanding, Zephyr NVS is the simplest and most lightweight storage compare to other file systems in Zephyr. It used for scenarios with limited storage needs and minimal complexity e.g., key-value storage, configurations . 
Briefly, NVS is the best choice due to its minimal footprint and ease of use.

Laczen commented:
@andrisk-dev could you change the PR start (just add ~ at start and end) reformulate so that it is just a change to avoid a static analysis flag. Also update your commit message as not to get any confusion. Once this is done I will approve the change.

andrisk-dev commented:
> @andrisk-dev could you change the PR start (just add ~ at start and end) reformulate so that it is just a change to avoid a static analysis flag. Also update your commit message as not to get any confusion. Once this is done I will approve the change.

OK, I made the requested changes to the PR.
